tests/trainer/logging_/test_eval_loop_logging.py

```diff
@@ -672,3 +673,29 @@ def val_dataloader(self):
         enable_model_summary=False,
     )
     trainer.fit(model)
+
+
+@pytest.mark.parametrize("num_dataloaders", [1, 2, 11])
+def test_log_metrics_only_include_metrics_from_concerned_dataloader(num_dataloaders, tmpdir):
+    class TestModel(BoringModel):
+        def test_step(self, batch, batch_idx, dataloader_idx=0):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            self.log("fake_test_acc", loss)
+            return {"y": loss}
+
+        def test_epoch_end(self, *_) -> None:
+            pass
+
+    test = RandomDataset(32, 2)
+    test_dataloaders = [DataLoader(test, batch_size=1)] * num_dataloaders
+
+    model = TestModel()
+    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1)
+
+    output = trainer.test(model, dataloaders=test_dataloaders)
+    assert sum(len(x) for x in output) == num_dataloaders
+    if num_dataloaders == 1:
+        assert "dataloader_idx" not in output[0]
+    else:
+        assert all(f"dataloader_idx_{idx}" == list(x.keys())[0].split("/")[1] for idx, x in enumerate(output))
```

Comment:
Actually, I think this could unit-test `_filter_metrics_for_dataloader` instead of run `trainer.test`

Reply:
Yeah you do have a valid point. It is just that coupling a unit test to the string prefix (which isn't really central to the logic) seems a tad too precise or may be that's the point.

Reply:
We don't make a good example of it but we are trying to unit-test more instead of having full trainer tests.

It's much faster CI-wise, less prone to flakiness, and it's easier to understand and validate different implementations.

So yes, let's change this to a unit-test

Reply:
Hi @shabie 

I've pushed a commit (51a9e64) replacing the existing test with the unit-tests that I was proposing.

With this, the runtime has gone from 124 ms to 1ms (in my machine).

It also helped uncover a bug, for which I've added a test case. Can you have a look?

Reply:
OK will check.